index.bs

```diff
@@ -4046,7 +4073,57 @@ To override this default policy and indicate that a cross-origin <{iframe}> is a
+
+This can make deployment challenging for large environments where multiple country-specific domains are in use (e.g. example.com vs example.co.uk vs example.sg), where alternative or brand domains are required (e.g. myexampletravel.com vs examplecruises.com), and/or where platform as a service providers are used to support mobile apps. 
+
+[=[WRPS]=] can opt in to allowing [=WebAuthn Clients=] to enable a credential to be created and used across a limited set of related [=origin|origins=].
+Such [=[RPS]=] MUST choose a common [=RP ID=] to use across all ceremonies from related origins.
+
+A JSON document MUST be hosted at the `webauthn` well-known URL ([[!RFC8615]]) for the [=RP ID=]. The JSON document MUST be returned as follows:
 
+    - The content type MUST be `application/json`.
+    - The top-level JSON object MUST contain a key named `origins` whose value MUST be an array of one or more strings containing web origins.
+
+For example, for the RP ID `example.com`:
+
+<xmp class="json" highlight="json">
+{
+    "origins": [
+        "https://example.co.uk",
+        "https://example.de",
+        "https://example.sg"
+        "https://myexamplerewards.com"
+    ]
+}
+</xmp>
+
+[=WebAuthn Clients=] supporting this feature MUST support at least five [=origin labels=]. Client policy SHOULD define an upper limit to prevent abuse.
+
+[=WebAuthn Clients=] supporting this feature SHOULD include {{ClientCapability/relatedOrigins}} in their response to [[#sctn-getClientCapabilities|getClientCapabilities()]].
+
+### Validating Related Origins ### {#sctn-validating-relation-origin}
+
+To validate the calling origin is an authorized related origin for a given ceremony:
+
+1. Let |maxLabels| be the maximum number of [=origin labels=] allowed by client policy.
+1. Fetch the well-known URL for the RP ID (<code>https://|rpIdRequested|/.well-known/webauthn</code>).
+    1. If the fetch fails, the response does not have a content type of `application/json`, or does not have a status code (after following redirects) of 200, then throw a "{{SecurityError}}" {{DOMException}}.
+    1. If the body of the resource is not a valid JSON object, then throw a "{{SecurityError}}" {{DOMException}}.
+    1. If the value of the |origins| property of the JSON object is missing, or is not an array of strings, then throw a "{{SecurityError}}" {{DOMException}}.
+1. Let |labelsSeen| be a new empty [=set=].
+1. [=set/For each=] string in |origins|:
+    1. Let |url| be the result of parsing the string as a URL. If that fails, [=continue=].
+    1. Let |domain| be the [=effective domain=] of |url|. If that is null, [=continue=].
+    1. Remove any [=public suffix=] from the end of |domain|, including private registries and unknown registries. If |domain| is now empty, [=continue=].
+    1. Split |domain| into |labels| and let |label| be the right-most one.
+    1. If |labelsSeen| [=set/contains=] |label|, [=continue=].
```

Comment:
If https://example.com/.well-known/webauthn served a list of related origins containing `["https://example.co.jp", "https://example.us"]`, this logic would not allow https://example.us to use `"example.com"` as an RP ID because the label `"example"` would end up in `labelsSeen` when processing `"https://www.example.co.jp"`. The next step wouldn't get to evaluate to return `TRUE`.

Reply:
```js
const CLIENT_MAX_LABELS = 5;

// Hosted from https://example.com/.well-known
const WELL_KNOWN_WEBAUTHN = {
  origins: [
    "https://example.co.jp",
    "https://example.us",
    "https://otherexample.com",
    "https://onemoreexample.com",
  ]
};

function checkOriginIsRelated(currentOrigin) {
  const labelsSeen = new Set();

  const originURL = new URL(currentOrigin);
  const originDomain = originURL.hostname;
  const parsedOrigin = psl.parse(originDomain);

  for (let origin of WELL_KNOWN_WEBAUTHN.origins) {
    // 1. Let url be the result of parsing the string as a URL. If that fails, continue.
    const url = new URL(origin);

    // 2. Let domain be the effective domain of url. If that is null, continue.
    const domain = url.hostname;

    // 3. Remove any public suffix from the end of domain, including private registries and unknown
    // registries. If domain is now empty, continue.
    const parsedDomain = psl.parse(domain);

    if (parsedDomain.error) continue;

    // 4. Split domain into labels and let label be the right-most one.
    const label = parsedDomain.sld;

    if (!label) continue;

    // 5. If labelsSeen contains label, continue.
    if (labelsSeen.has(label)) continue;

    // 6. If callerOrigin's effective domain and domain are same origin, return True.
    if (parsedDomain.domain === parsedOrigin.domain) return true;

    // 7. Append label to labelsSeen.
    labelsSeen.add(label);

    // 8. If the size of labelsSeen is greater than or equal to maxLabels, break.
    if (labelsSeen.size >= CLIENT_MAX_LABELS) break;
  }

  return false;
}

// This is the second origin in the list, should be `true`
console.log(checkOriginIsRelated('https://example.us')) // false
```

Reply:
I'm trying to think of the best way to fix this and right now I think removing this step completely might be it.

```suggestion
```

Keeping the step to add the label to `labelsSeen` after the domain check still seems valid, as does checking the size of `labelsSeen` afterwards to make sure it conforms to the client's maximum number of allowed labels.

But now I'm questioning the value of the limiting the number of allowed labels in this way. A list of origins like this would also fail to understand a related origin, despite the fact that the last item in the list of origins should be valid:

```
[
  "https://example2.com",
  "https://example3.com",
  "https://example4.com",
  "https://example5.com",
  "https://example6.com",
  "https://example2.net",
]
```

`"example6"` would be the fifth label so the loop would be terminated. However `"https://example2.net" should get evaluated because its label, `"example2"` is still within the five-label minimum allowed by this capability.

Reply:
I just realized that `"example2.com"`, `"example2.net"`, and `"www.example2.us"` all have the same label `"example2"`. In a long enough list of related origins they may not get evaluated if the maximum number of labels gets hit before they're evaluated.

To what degree could we ask clients to lexicographically sort such a list of related origins by their label? It would remove some implementation burden from the RP, and prevent weird edge cases like this from complicating this logic further.

Reply:
Alternatively if we wanted to maintain the implementation friction we could add in a step for RPs to perform the lexicographical sort manually when maintaining their list of related origins.

Reply:
addressed in https://github.com/w3c/webauthn/pull/2040/commits/a108a2c79e081e78d68646de08cf78dafd2d0f21